### PR TITLE
List active measurement services in interactive approach

### PR DESCRIPTION
**Pull request by Jotheeswaran-Nandagopal**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

- Implement CLI command for interactive approach mode of creating measurement client template.
- Implement a method to list all the active measurement services and get the user input for the name of the module to be generated.

- The error handling and corner cases are addressed for the listing the active measurements, and handling error for the module name and other corner cases for template creation are planned to address in next pull request.

### Why should this Pull Request be merged?

- This implementation supports the Interactive Approach for the [Measurement Plug-In client generation](https://dev.azure.com/ni/DevCentral/_git/ASW?path=/Specs/Test%20Automation%20Frameworks/E2719769_Support%20for%20Measurement%20Plug-In%20client%20APIs/E2719769_Support%20for%20Measurement%20Plug-In%20client%20APIs.md&version=GBmain&_a=preview&anchor=interactive-approach)'s in python.

### What testing has been done?

Existing tests passed and manually tested the new implementation

- Planned to implement acceptance test for the Measurement client generator, after completing the implementation of the template generation.

### CODE REVIEW

File: packages/generator/ni_measurement_plugin_client_generator/template.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import click
+from ni_measurement_plugin_sdk_service.discovery import DiscoveryClient
+from ni_measurement_plugin_sdk_service.measurement.info import ServiceInfo
+
+
+_V2_MEASUREMENT_SERVICE_INTERFACE = "ni.measurementlink.measurement.v2.MeasurementService"
+
+
+def _get_module_name_and_service_class(
+    available_services: Sequence[ServiceInfo],
```
Comment (SowmyaRavee) [inline packages/generator/ni_measurement_plugin_client_generator/template.py:14]:
```suggestion
    registered_services: Sequence[ServiceInfo],
```

File: packages/generator/ni_measurement_plugin_client_generator/template.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+) -> Tuple[str, str]:
+    print("List of registered measurements: ")
+    for i, service in enumerate(available_services):
+        print(f"{i+1}. {service.display_name}")
+
+    try:
+        index = int(
+            input("\nEnter the serial number for the desired measurement plugin client creation: ")
```
Comment (SowmyaRavee) [inline packages/generator/ni_measurement_plugin_client_generator/template.py:22]:
```suggestion
            input("\nEnter the serial number to create the desired measurement plugin client: ")
```

Comment (SowmyaRavee) [reply packages/generator/ni_measurement_plugin_client_generator/template.py:22]: The text needs update as we support multiple measurement migration

File: packages/generator/ni_measurement_plugin_client_generator/template.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+def _get_module_name_and_service_class(
+    available_services: Sequence[ServiceInfo],
+) -> Tuple[str, str]:
+    print("List of registered measurements: ")
+    for i, service in enumerate(available_services):
+        print(f"{i+1}. {service.display_name}")
+
+    try:
+        index = int(
```
Comment (SowmyaRavee) [inline packages/generator/ni_measurement_plugin_client_generator/template.py:21]:
```suggestion
        selected_measurement_index = int(
```

File: packages/generator/ni_measurement_plugin_client_generator/template.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    return (module_name, measurement_service_class)
+
+
+@click.command()
+@click.argument("module_name", type=str, default="")
+@click.option(
+    "-i",
+    "--interactive-mode",
+    is_flag=True,
+    help="Utilize interactive input for Python measurement client creation.",
```
Comment (SowmyaRavee) [inline packages/generator/ni_measurement_plugin_client_generator/template.py:41]:
```suggestion
    help="Interactive mode for Python measurement client creation.",
```

File: packages/generator/ni_measurement_plugin_client_generator/template.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    You can use the generated module to interact with the corresponding measurement service.
+
+    MODULE_NAME: Name for the Python measurement client module to be generated.
+    """
+    discovery_client = DiscoveryClient()
+    available_services = discovery_client.enumerate_services(_V2_MEASUREMENT_SERVICE_INTERFACE)
+
+    if not available_services:
+        print("Could find any active measurements. Please start one and try again.")
```
Comment (SowmyaRavee) [inline packages/generator/ni_measurement_plugin_client_generator/template.py:54]:
```suggestion
        print("No active measurements were found. Please start one and try again.")
```

**Comment by SowmyaRavee** [comment]
Please update the description `Implement CLI command for interactive approach mode of creating measurement client template.`

I feel we can have the PR created after completing the `create_client` method as the method is incomplete now

File: packages/generator/ni_measurement_plugin_client_generator/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+"""Measurement Plugin client Code Generator."""
```
Comment (jayaseelan-james) [inline packages/generator/ni_measurement_plugin_client_generator/__init__.py:1]:
```suggestion
"""Measurement Plugin Client Code Generator."""
```

File: packages/generator/ni_measurement_plugin_client_generator/__init__.py
Comment (jayaseelan-james) [inline packages/generator/ni_measurement_plugin_client_generator/__init__.py:1]: Should the directory be named `ni_measurement_plugin_sdk_client_generator` to be consistent with the existing directories?

File: packages/generator/ni_measurement_plugin_client_generator/template.py (hunk 4 above)
Comment (jayaseelan-james) [reply packages/generator/ni_measurement_plugin_client_generator/template.py:41]: Should we say "... measurement plug-in client" here?

File: packages/generator/ni_measurement_plugin_client_generator/template.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+)
+def create_client(module_name: str, interactive_mode: bool) -> None:
+    """Generates a Python measurement client module for a measurement service from the template.
+
+    You can use the generated module to interact with the corresponding measurement service.
+
+    MODULE_NAME: Name for the Python measurement client module to be generated.
+    """
+    discovery_client = DiscoveryClient()
+    available_services = discovery_client.enumerate_services(_V2_MEASUREMENT_SERVICE_INTERFACE)
```
Comment (jayaseelan-james) [inline packages/generator/ni_measurement_plugin_client_generator/template.py:51]: Can this be `registered_measurement_services`?

File: packages/generator/ni_measurement_plugin_client_generator/template.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+import click
+from ni_measurement_plugin_sdk_service.discovery import DiscoveryClient
+from ni_measurement_plugin_sdk_service.measurement.info import ServiceInfo
+
+
+_V2_MEASUREMENT_SERVICE_INTERFACE = "ni.measurementlink.measurement.v2.MeasurementService"
+
+
+def _get_module_name_and_service_class(
```
Comment (jayaseelan-james) [inline packages/generator/ni_measurement_plugin_client_generator/template.py:13]: Would it make sense to split this method to be more granular? Where one method will take care of displaying the registered services and another method to get the input index?

By this change, I guess we can call both methods in the `create_client` itself.

**Comment by Jotheeswaran-Nandagopal** [comment]
Team, as this implementation is incomplete as looking as template creation. I will close this pull request and create a new one with measurement plugin client creation implementation.

**Pull request closed without merging**